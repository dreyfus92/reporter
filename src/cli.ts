#!/usr/bin/env node

import fs from 'node:fs/promises';
import {createRequire} from 'node:module';
import sade from 'sade';
import c from 'picocolors';
import {report} from './index.js';
import type {Message} from './types.js';

const version = createRequire(import.meta.url)('../package.json').version;
const cli = sade('e18e', false)
  .version(version)
  .option(
    '--pack',
    `Package manager to use for packing ('auto' | 'npm' | 'yarn' | 'pnpm' | 'bun')`,
    'auto'
  );

cli.command('run [root]', '', {default: true}).action(async (root, opts) => {
  console.log(`${c.bold(`@e18e/reporter v${version}`)}`);
  console.log('-'.repeat(21));
  console.log();

  // If a path is passed, see if it's a path to a file (likely the tarball file)
  if (root) {
    const stat = await fs.stat(root).catch(() => {});
    const isTarballFilePassed = !!stat?.isFile();
    if (isTarballFilePassed) {
      opts.pack = {tarball: (await fs.readFile(root)).buffer};
    }
  }

  const {info, messages} = await report({root, pack: opts.pack});

  console.log(c.cyan('Package info'));
  console.log(`${c.dim('Name   ')}  ${info.name}`);
  console.log(`${c.dim('Version')}  ${info.version}`);
  console.log(`${c.dim('Type   ')}  ${info.type.toUpperCase()}`);
  console.log();

  console.log(c.cyan('Package report'));
  if (messages.length === 0) {
    console.log(c.bold(c.green('All good!')));
  } else {
    formatMessages(messages).forEach((l) => console.log(l));
  }
  console.log();
});

cli.parse(process.argv);

function formatMessages(messages: Message[]) {
  /** @type {string[]} */
  const logs = [];

  const errors = messages.filter((v) => v.severity === 'error');
  if (errors.length) {
    logs.push(c.bold(c.red('Errors:')));
    errors.forEach((m, i) => logs.push(c.dim(`${i + 1}. `) + m.message));
    process.exitCode = 1;
  }

  const warnings = messages.filter((v) => v.severity === 'warning');
  if (warnings.length) {
    logs.push(c.bold(c.yellow('Warnings:')));
    warnings.forEach((m, i) => logs.push(c.dim(`${i + 1}. `) + m.message));
  }

  const suggestions = messages.filter((v) => v.severity === 'suggestion');
  if (suggestions.length) {
    logs.push(c.bold(c.blue('Suggestions:')));
    suggestions.forEach((m, i) => logs.push(c.dim(`${i + 1}. `) + m.message));
  }

  return logs;
}
