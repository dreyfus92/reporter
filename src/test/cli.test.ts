import {vi, describe, it, expect, beforeEach, afterEach} from 'vitest';
import {runCli} from '../cli.js';

// Mock the browser environment
vi.mock('../analyze-dependencies-browser.js', () => ({
  LocalDependencyAnalyzer: class MockLocalAnalyzer {
    async analyzeDependencies() {
      throw new Error('Local dependency analysis is not supported in the browser');
    }
  },
  RemoteDependencyAnalyzer: class MockRemoteAnalyzer {
    async analyzeDependencies() {
      throw new Error('Remote dependency analysis is not supported in the browser');
    }
  },
  analyzeDependencies: async () => {
    throw new Error('Local dependency analysis is not supported in the browser');
  }
}));

describe('CLI in browser environment', () => {
  const originalWindow = global.window;

  beforeEach(() => {
    // Mock browser environment
    (global as any).window = { document: {} };
  });

  afterEach(() => {
    // Restore window object
    (global as any).window = originalWindow;
  });

  it('should throw an error when run in a browser environment', async () => {
    await expect(runCli(['--pack', 'npm'])).rejects.toThrow(
      'Local dependency analysis is not supported in the browser'
    );
  });
}); 