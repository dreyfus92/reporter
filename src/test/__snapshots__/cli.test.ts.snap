// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI > should display package report 1`] = `
"Generate a performance report for your package. (e18e-report v0.0.1)

┌  Generating report...
│
●  Tarball Analysis
│  Total deps      1
│  Direct deps     1
│  Dev deps        0
│  CJS deps        N/A
│  ESM deps        N/A
│  Install size    170.0 B
│
│  Dependency type analysis is only available for local analysis, as tarballs do not include dependencies.
│
●  Package report
│
│  Suggestions:
│    • No type definitions found.
│
│
└  Report generated successfully!

"
`;

exports[`CLI > should display package report 2`] = `""`;
